Add missing preview features of Deployment and Deployment Statuses API

pascal-hofmann opened the pull request:
- Enable support for deployment status states `in_progress` and `queued`
  - Add arguments `production_environment` and `transient_environment` to `Repository.create_deployment()`
  - Add attributes `production_environment` and `transient_environment` to `Deployment`
  - Fix incorrect assertion for `description` in `Deployment.create_status()`
  - Add arguments `environment`, `environment_url`, `auto_inactive` to` Deployment.create_status()`
  - Add attribute `environment_url` to `DeploymentStatus`


References:
  - https://docs.github.com/en/rest/reference/repos#deployments
  - https://developer.github.com/changes/2016-04-06-deployment-and-deployment-status-enhancements/
  - https://developer.github.com/changes/2018-10-16-deployments-environments-states-and-auto-inactive-updates/

s-t-e-v-e-n-k commented on github/Deployment.py:
```diff
@@ -71,7 +88,7 @@ def task(self):
     def payload(self):
         """
-        :type: dict
+        :type: string
```
This is an API break.

pascal-hofmann replied:
If I remember correctly the payload returned already was a string, even if stated otherwise here. I'll have another look at this.

pascal-hofmann replied:
Github just returns whatever is passed. No matter which type. I've changed the assert in create_deployment to dict and reverted my changes.

s-t-e-v-e-n-k commented on github/ApplicationOAuth.pyi:
```diff
@@ -12,13 +12,6 @@ class ApplicationOAuth(NonCompletableGithubObject):
-        redirect_uri: Optional[str],
-        state: Optional[str],
-        login: Optional[str]
+        self, redirect_uri: Optional[str], state: Optional[str], login: Optional[str]
```
Why are you making these changes?

pascal-hofmann replied:
Tox made these. I just ran it.

s-t-e-v-e-n-k commented on github/Branch.py:
```diff
@@ -163,7 +163,7 @@ def edit_protection(
             required_approving_review_count is github.GithubObject.NotSet
             or isinstance(required_approving_review_count, int)
-        ), required_approving_review_count
+        ), (required_approving_review_count)
```
As above -- why are you making these changes?

pascal-hofmann replied:
Tox made these. I just ran it. I'll amend my commit and not include these changes.

s-t-e-v-e-n-k reviewed (changes requested):
There is a awful lot more here changed than what the PR says, would you mind cleaning this up to just doing that to make this easier to review?

pascal-hofmann commented:
@s-t-e-v-e-n-k I've removed the linted files. Now the build fails again: https://github.com/PyGithub/PyGithub/pull/1674/checks?check_run_id=1049885210

I guess this should be fixed on master?

s-t-e-v-e-n-k commented:
```
lint run-test: commands[0] | pre-commit run --all-files --show-diff-on-failure
black....................................................................Passed
```

is what I get locally, so I'm still digging into why CI has no love.

pascal-hofmann commented:
@s-t-e-v-e-n-k I think the CI issue is independent from my PR (other PRs have the same problem). Can this be merged or are there still things I should improve?

s-t-e-v-e-n-k commented:
I haven't had time yet to dig into the failure yet, I hope to get some time tomorrow.

pascal-hofmann commented:
@s-t-e-v-e-n-k I rebased this on latest master. Now all the checks pass.

pascal-hofmann commented:
@s-t-e-v-e-n-k It would be great to have this merged and packed in a new release of PyGit.

Is there anything I can do to help with this?

pascal-hofmann commented:
I just noticed that the commit message was outdated. This is now fixed.

pascal-hofmann commented:
@adamtheturtle @jzelinskie @nhomar @jacquev6 @sfdye @s-t-e-v-e-n-k 

It would be great to get this merged and released. I noticed that there is not much of maintainer activity lately and quite some outstanding contributions. I volunteer as maintainer. This is a great project, I want to move it forward.

sfdye commented:
@pascal-hofmann Thanks for your proposal, let's talk in private.

s-t-e-v-e-n-k commented on github/Deployment.py:
```diff
@@ -150,6 +167,14 @@ def get_statuses(self):
             self._requester,
             self.url + "/statuses",
             None,
+            headers={
```
This is repeated a fair amount -- would you consider a static method in this class that returns the string?

pascal-hofmann replied:
This is now implemented.

s-t-e-v-e-n-k commented on github/Repository.py:
```diff
@@ -1836,15 +1840,20 @@ def create_deployment(
         :param: auto_merge: bool
-        :param: required_contexts: list of statuses
+        :param: required_contexts: list of status contexts
         :param: payload: json
```
This is no longer json

s-t-e-v-e-n-k replied:
Further more, we need to remember to mention the API break on the next release.

pascal-hofmann replied:
This is now implemented.

s-t-e-v-e-n-k reviewed (changes requested):
This is looking so close, thanks for your changes. Some doc changes, and you're missing typing hint changes to Repository.pyi and I think we can finally merge this!

pascal-hofmann commented:
Hi @s-t-e-v-e-n-k,
sorry for the delay. I updated the stub files, fixed the payload type and rebased on master. I also moved code for `production_environment()` and `transient_environment()` to a place within `github/Deployment.py` where it makes more sense.

I did a separate commit with these changes for now, so you can easily see latest changes. Give me a shout if I should squash the commits.

s-t-e-v-e-n-k commented:
Apparently, we have the technology to squash commits at merge time. :-)

s-t-e-v-e-n-k commented on github/Repository.pyi:
```diff
@@ -295,6 +308,14 @@ class Repository(CompletableGithubObject):
     def get_contributors(
         self, anon: Union[str, _NotSetType] = ...
     ) -> PaginatedList[NamedUser]: ...
+    def get_deployment(self, id_: int) -> Any: ...
```
I mean, surely this is Deployment, not Any. However, I don't want to hold this PR any further.

pascal-hofmann replied:
Good catch. Fixed in  https://github.com/PyGithub/PyGithub/pull/1767/

pascal-hofmann commented:
@s-t-e-v-e-n-k Thanks for all your work on this PR and this project in general!
